File: 3.cameraandmatrices/mobile/Engine/Source/Objects/CCTile3DButton.cpp
/*-----------------------------------------------------------
 * 2c - Cross Platform 3D Application Framework
 *-----------------------------------------------------------
 * Copyright © 2010 – 2011 France Telecom
 * This software is distributed under the Apache 2.0 license,
 * see the "license.txt" file for more details.
 *-----------------------------------------------------------
 * File Name   : CCTile3DButton.cpp
 *-----------------------------------------------------------
 */

#include "CCDefines.h"
#include "CCObjects.h"
#include "CCTextureFontPage.h"


CCTile3DButton::CCTile3DButton(CCSceneBase *scene)
{
    construct( scene );
}


CCTile3DButton::CCTile3DButton(CCSceneBase *scene, const float width, const float height, const char *text)
{
    construct( scene );
    
    setupBase( width, height );
    
    if( text )
    {
        textModel->setText( text, height );
    }
}


CCTile3DButton::CCTile3DButton(CCSceneBase *scene, const float width, const char *textureName)
{
    construct( scene );
    
    const CCTextureBase *texture = gEngine->textureManager->getTexture( textureName, Resource_Packaged );
    const float aspectRatio = texture->getImageWidth() / texture->getImageHeight();
    const float height = width / aspectRatio;
    setupBase( width, height );

    setBaseTexture( textureName, Resource_Packaged );
}


CCTile3DButton::CCTile3DButton(CCSceneBase *scene, const char *text, const float height, const bool centered)
{
    construct( scene );
    
    textModel->setText( text, height );
    
    const float width = textModel->getWidth();
    
    setHCollisionBounds( width * 0.5f, height * 0.5f, CC_SMALLFLOAT );
    
    textModel->setCentered( centered );
    
    if( centered == false )
    {
        translate( collisionBounds.x, 0.0f, 0.0f );
    }
}


void CCTile3DButton::construct(CCSceneBase *scene)
{
    AddFlag( collideableType, collision_ui );

    if( scene != NULL )
    {
        setScene( scene );
    }
    
    renderDepth = false;
    setTransparent();
    
    model = new CCModelBase();
    
    baseModel = new CCModelBase();
    model->addModel( baseModel );
    setBaseColour( CCColour() );
    colourInterpolator.setDuration( 0.5f );
    baseSquare = NULL;
    
    textModel = new CCModelText( this );
    
    if( scale == NULL )
    {
        CCVector3FillPtr( &scale, 0.0f, 0.0f, 1.0f );
    }
    scaleInterpolator.setup( scale, 1.0f );

    CCUpdateCollisions( this );
}


void CCTile3DButton::destruct()
{
    super::destruct();
}

// CCRenderable
void CCTile3DButton::refreshModelMatrix()
{
	if( updateModelMatrix )
	{
        CCMatrixLoadIdentity( modelMatrix );	
        CCMatrixTranslate( modelMatrix, position.x, position.y, position.z );
        
		if( scale != NULL )
		{	
			CCMatrixScale( modelMatrix, scale->x, scale->y, scale->z );
		}

        if( rotation.x != 0.0f )
        {
            CCMatrixRotate( modelMatrix, rotation.x, 1.0f, 0.0f, 0.0f );
        }

        if( rotation.y != 0.0f )
        {
            CCMatrixRotate( modelMatrix, rotation.y, 0.0f, 1.0f, 0.0f );
        }
        
		updateModelMatrix = false;
		dirtyWorldMatrix();
	}
}


// CCSceneObject
void CCTile3DButton::update(const CCTime &gameTime)
{
    super::update( gameTime );

    if( scale != NULL )
    {
        const float speed = gameTime.delta * 1.0f;
        if( scaleInterpolator.update( speed ) )
        {
            dirtyModelMatrix();
        }
        else
        {
            DELETE_POINTER( scale );
        }
    }

    colourInterpolator.update( gameTime.delta );

    if( textModel != NULL )
    {
        textModel->colourInterpolator.update( gameTime.delta );
    }
}


void CCTile3DButton::renderModels(const bool alpha)
{
    if( alpha )
    {
        if( renderDepth )
        {
            glEnable( GL_DEPTH_TEST );
            super::renderModels( alpha );
            glDisable( GL_DEPTH_TEST );
        }
        else
        {
            super::renderModels( alpha );
        }
    }
    else
    {
        if( renderDepth )
        {
            super::renderModels( alpha );
        }
        else
        {
            glDisable( GL_DEPTH_TEST );
            super::renderModels( alpha );
            glEnable( GL_DEPTH_TEST );
            
        }
    }
}


void CCTile3DButton::setupBase(const float width, const float height)
{
    if( baseSquare == NULL )
    {
        baseSquare = new CCPrimitiveSquare();
        baseModel->addPrimitive( baseSquare );
    }

    const float hWidth = width * 0.5f;
    const float hHeight = height * 0.5f;
    setHCollisionBounds( hWidth, hHeight, CC_SMALLFLOAT );
    baseSquare->setupZFacing( collisionBounds.x, collisionBounds.y );
    CCUpdateCollisions( this );
}


void CCTile3DButton::setBaseTexture(const char *name, const CCResourceType resourceType)
{
    baseSquare->setTexture( name, resourceType );
}


void CCTile3DButton::setBaseColour(const CCColour &inColour)
{
    baseModel->setColour( inColour );
    colourInterpolator.setup( baseModel->getColour(), *baseModel->getColour() );
}


void CCTile3DButton::setBaseColourAlpha(const float inAlpha)
{
    baseModel->setColourAlpha( inAlpha );
    colourInterpolator.setup( baseModel->getColour(), *baseModel->getColour() );
}

File: 2.loadingtextures/webgl/source/js/engine/objects/CCTile3DButton.js
/*-----------------------------------------------------------
 * 2c - Cross Platform 3D Application Framework
 *-----------------------------------------------------------
 * Copyright © 2011 - 2011 France Telecom
 * This software is distributed under the Apache 2.0 license.
 * http://www.apache.org/licenses/LICENSE-2.0.html
 *-----------------------------------------------------------
 * File Name   : CCTile3DButton.js
 * Description : A dynamic button widget.
 *
 * Created     : 09/08/11
 * Author(s)   : Ashraf Samy Hegab
 *-----------------------------------------------------------
 */

function CCTile3DButton(scene)
{
    this.construct( scene );
}
CopyPrototype( CCTile3DButton, CCTile3D, "CCTile3D" );


CCTile3DButton.prototype.construct = function(scene)
{
	this.CCTile3D_construct();
	
	if( scene )
	{
		this.setScene( scene );
	}
	
	this.renderDepth = false;
	this.setTransparent( true );

    var model = this.model = new CCModelBase();
    
    var baseModel = this.baseModel = new CCModelBase();
    model.addModel( baseModel );
	this.colourInterpolator = new CCInterpolatorColour();
	this.colourInterpolator.setDuration( 0.5 );
	this.setBaseColour( gColour.white() );
	this.baseSquare = false;
	
	this.textModel = new CCModelText( this );
}


CCTile3DButton.prototype.setup = function(width, height, text)
{
	this.setupBase( width, height );
    
    if( text )
    {
        this.textModel.setText( text, height );
    }
	
	return this;
}

// Create tile with width and textures aspect ratio
CCTile3DButton.prototype.setupTexture = function(width, textureSrc)
{
	var TextureLoadCallback = function(tile, src) 
	{
		this.tile = tile;
		this.src = src;
	}
    TextureLoadCallback.prototype.callback = function(texture)
    {
        var aspectRatio = texture.width / texture.height;
		var height = width / aspectRatio;
		this.tile.setupBase( width, height );
		
		this.tile.setBaseTexture( this.src );
    }
	
	gEngine.textureManager.getTexture( textureSrc, new TextureLoadCallback( this, textureSrc ) );
	return this;
}

// Create tile with the text height
CCTile3DButton.prototype.setupText = function(text, height, centered)
{
	var textModel = this.textModel;
	textModel.setText( text, height );
    
    var width = textModel.getWidth();
    this.setHCollisionBounds( width * 0.5, height * 0.5, CC_SMALLFLOAT );
    
    textModel.setCentered( centered );
    
    if( centered == false )
    {
        this.translate( collisionBounds.x, 0.0, 0.0 );
    }
	
	return this;
}


CCTile3DButton.prototype.destruct = function(width, height)
{
}


CCTile3DButton.prototype.update = function(delta)
{
	this.colourInterpolator.update( delta );
	
	if( this.textModel != false )
    {
        this.textModel.colourInterpolator.update( delta );
    }
}


CCTile3DButton.prototype.setupBase = function(width, height)
{
    if( this.baseSquare == false )
    {
        var baseSquare = this.baseSquare = new CCPrimitiveSquare();
        this.baseModel.addPrimitive( baseSquare );
    }

    var hWidth = width * 0.5;
    var hHeight = height * 0.5;
    this.setHCollisionBounds( hWidth, hHeight, CC_SMALLFLOAT );
	this.baseSquare.setScale( width, height, 1.0 );
    
    this.refreshCollisions();
}


CCTile3DButton.prototype.setBaseTexture = function(src)
{
	this.baseSquare.setTexture( src );
}


CCTile3DButton.prototype.setBaseColour = function(inColour)
{
	this.baseModel.setColour( inColour );
	this.colourInterpolator.setup( this.baseModel.colour );
}